components/org.wso2.carbon.identity.sso.agent/src/main/java/org/wso2/carbon/identity/sso/agent/saml/artifact/SAMLSSOArtifactResolutionService.java

```diff
@@ -0,0 +1,259 @@
+
+        this.ssoAgentConfig = ssoAgentConfig;
+    }
+
+    /**
+     * Create a SAML artifactResolveObject based on given samlArt parameter and wrapped into a soapRequest
+     * Send the soapRequest to the Artifact Resolve Endpoint
+     *
+     * @param samlArtReceived SAML Artifact reference needed to get the actual data
+     * @return ArtifactResponse
+     */
+    public ArtifactResponse getSAMLArtifactResponse(String samlArtReceived) throws ArtifactResolutionException {
+
+        validateArtifactResolveConfig();
+        ArtifactResolve artifactResolve = generateArtifactResolveReq(samlArtReceived);
+        return sendArtifactResolveRequest(artifactResolve);
+    }
+
+    /**
+     * Create SAML ArtifactResolve Object and sign
+     *
+     * @param samlArtReceived object reference to actual data
+     * @return SAML ArtifactResolve Object
+     */
+    public ArtifactResolve generateArtifactResolveReq(String samlArtReceived) throws ArtifactResolutionException {
+
+        ArtifactResolve artifactResolve = createArtifactResolveObject(samlArtReceived);
+        if (ssoAgentConfig.getSAML2().isEnableArtifactResolveSigning()) {
+            if (log.isDebugEnabled()) {
+                log.debug("Signing artifact resolve request for the received SAML artifact: " + samlArtReceived);
+            }
+            artifactResolve = signArtifactResolveReq(artifactResolve);
+        }
+        return artifactResolve;
+    }
+
+    /**
+     * Send Artifact Resolve Request to Artifact Resolution Service
+     *
+     * @param artifactResolve Artifact Resolve Request
+     * @return ArtifactResponse
+     * @throws ArtifactResolutionException
+     */
+    public ArtifactResponse sendArtifactResolveRequest(ArtifactResolve artifactResolve)
+            throws ArtifactResolutionException {
+
+        SAMLSSOSoapMessageService soapMessageService = new SAMLSSOSoapMessageService();
+        Envelope envelope = soapMessageService.buildSOAPMessage(artifactResolve);
+        String envelopeElement;
+        try {
+            envelopeElement = SSOAgentUtils.marshall(envelope);
+        } catch (SSOAgentException e) {
+            throw new ArtifactResolutionException("Encountered error marshalling SOAP message with artifact " +
+                    "resolve, into its DOM representation", e);
+        }
+
+        if (log.isDebugEnabled()) {
+            log.debug("Artifact Resolve Request as a SOAP Message: " + envelopeElement);
+        }
+
+        String artifactResponseString = soapMessageService.sendSOAP(envelopeElement,
+                ssoAgentConfig.getSAML2().getArtifactResolveURL());
+        ArtifactResponse artifactResponse = extractArtifactResponse(artifactResponseString);
+
+        if (artifactResponse != null && isArtifactResponseValid(artifactResolve, artifactResponse)) {
+            return artifactResponse;
+        } else {
+            throw new ArtifactResolutionException("Artifact Response is not valid.");
```

Comment:
Don't we miss some information here ?. When this exception is thrown to upper level we don't have a clue on what actually the response was. Isn't that a good idea to log the response somewhere ?